src/main/java/com/google/sps/servlets/IntervieweeFeedbackServlet.java

```diff
@@ -0,0 +1,147 @@
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.DatastorePersonDao;
+import com.google.sps.data.DatastoreScheduledInterviewDao;
+import com.google.sps.data.EmailSender;
+import com.google.sps.data.Person;
+import com.google.sps.data.PersonDao;
+import com.google.sps.data.ScheduledInterview;
+import com.google.sps.data.ScheduledInterviewDao;
+import com.google.sps.data.TimeRange;
+import com.sendgrid.Response;
+import com.sendgrid.SendGrid;
+import com.sendgrid.helpers.mail.Mail;
+import com.sendgrid.helpers.mail.objects.Content;
+import com.sendgrid.helpers.mail.objects.Email;
+import java.nio.file.Paths;
+import java.nio.file.Path;
+import java.io.IOException;
+import java.time.format.DateTimeFormatter;
+import java.time.format.TextStyle;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.time.ZoneId;
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletException;
+
+// Servlet that gets the feedback from an interviwer and sends it to an interviewee.
+@WebServlet("/interviewee-feedback")
+public class IntervieweeFeedbackServlet extends HttpServlet {
+  private ScheduledInterviewDao scheduledInterviewDao;
+  private PersonDao personDao;
+  private Path emailsPath =
+      Paths.get(
+          System.getProperty("user.home") + "/InterviewMe/src/main/resources/templates/email");
+
+  @Override
+  public void init() {
+    init(new DatastoreScheduledInterviewDao(), new DatastorePersonDao());
+  }
+
+  public void init(ScheduledInterviewDao scheduledInterviewDao, PersonDao personDao) {
+    this.scheduledInterviewDao = scheduledInterviewDao;
+    this.personDao = personDao;
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long scheduledInterviewId = Long.parseLong(request.getParameter("interviewId"));
+    HashMap<String, String> answers = new HashMap<String, String>();
+    for (int i = 1; i < 12; i++) {
+      String template = String.format("{{question_%s}}", i);
+      String param = String.format("question%s", i);
+      answers.put(template, request.getParameter(param));
+    }
+
+    String userEmail = UserServiceFactory.getUserService().getCurrentUser().getEmail();
+    String userId = UserServiceFactory.getUserService().getCurrentUser().getUserId();
+    // Since Users returned from the LocalUserService (in tests) do not have userIds, here we set
+    // the userId equal to a hashcode.
+    if (userId == null) {
+      userId = String.format("%d", userEmail.hashCode());
+    }
+
+    if (interviewExists(scheduledInterviewId)) {
```

Comment:
I recommend against this structure where you're looking up the scheduledinterview more than once. The lookups are probably the most time intensive part of this function, and you're doing it multiple times.

Consider instead updating these helper functions to accept a ScheduledInterview that you lookup in advance.

Reply:
Done